[C] Diagnose use of C++ keywords in C

AaronBallman opened the pull request:
This adds a new diagnostic group, -Wc++-keyword, which is off by default and grouped under -Wc++-compat. The diagnostic catches use of C++ keywords in C code.

This change additionally fixes an issue with -Wreserved-identifier not diagnosing use of reserved identifiers in function parameter lists in a function declaration which is not a definition.

Fixes https://github.com/llvm/llvm-project/issues/21898

AaronBallman committed:
[C] Diagnose use of C++ keywords in C

This adds a new diagnostic group, -Widentifier-is-c++-keyword, which is
off by default and grouped under -Wc++-compat. The diagnostic catches
use of C++ keywords in C code.

This change additionally fixes an issue with -Wreserved-identifier not
diagnosing use of reserved identifiers in function parameter lists.

Partially fixes #21898

AaronBallman committed:
Fix the parameter handling

AaronBallman committed:
Update comment; NFC

AaronBallman commented:
Note, I'm running this one through llvm-compile-time-tracker because of the lookup happening on every identifier. I'm not convinced that the "is this identifier a keyword in other language modes" code has good performance and wanted to double-check.

AaronBallman committed:
Fix formatting; NFC

AaronBallman commented:
> Maybe `-Wkeyword-in-c++` or `-Wc++-keyword` would be a more concise name for the group?

I'm not tied to the name I picked, so either of these is fine by me. GCC doesn't split this into its own warning group, so we've got some latitude.

Any strong preferences?

philnik777 commented:
> > Maybe `-Wkeyword-in-c++` or `-Wc++-keyword` would be a more concise name for the group?
> 
> I'm not tied to the name I picked, so either of these is fine by me. GCC doesn't split this into its own warning group, so we've got some latitude.
> 
> Any strong preferences?

FWIW I feel like the current one feels quite artificial. I think `-Wc++-keyword` is kinda nice, since it feels very similar to `-Wc++-compat`, but I don't have strong preferences.

shafik commented on clang/test/Sema/c++-keyword-in-c.c:
```diff
@@ -0,0 +1,213 @@
+
+// These are conditionally a keyword in C++, so they're intentionally not being
+// diagnosed as a keyword.
+int module;
```
what about `override` and `final`?

AaronBallman replied:
I'll add them as test cases; they're not intended to be diagnosed either.

Sirraide commented:
> I think `-Wc++-keyword` is kinda nice

+1

Sirraide reviewed (commented):
What about `and`, `or`, and friends (which I think are only keywords in C if you include a certain header irrc).

AaronBallman commented:
Alas, I think this PR may be a non-starter due to performance: https://llvm-compile-time-tracker.com/?config=Overview&stat=wall-time&remote=AaronBallman:
```
C 		56a3f3cd28 	8.04s (+2.96%) 	8.23s (+3.32%) 	11.87s (+1.92%) 	2.83s (+10.21%) 	8.50s (+2.99%) 	3.22s (+7.82%) 	7.12s (+1.63%) 	2.43s (+5.91%) 	580.64s (+0.44%)
```
+10% compile time overhead is painful. I believe the issue is with needing to look at token *names* (doing a string compare in a loop across a pile of tokens on every identifier declares). But the token kind we have is `identifier` and not one of the keywords, so we can't do the more trivial integer comparisons. Even if we could, I suspect this would have a performance hit just due to the number of identifiers.

I'm going to try checking whether the diagnostic is enabled before doing the comparisons to see if that improves things. The diagnostic is off by default and only enabled under `-Wc++-compat`, so some amount of performance hit is reasonable given the need to avoid C++ keywords in C code.

AaronBallman commented:
> What about `and`, `or`, and friends (which I think are only keywords in C if you include a certain header irrc).

Good call, I'll add those.

AaronBallman committed:
Only perform the check if the diagnostic is enabled

Checking to see how much performance this claws back

AaronBallman committed:
Add test cases for override and final

Sirraide commented:
> But the token kind we have is `identifier` and not one of the keywords, so we can't do the more trivial integer comparisons.

Can we precompute a list/map/whatever of all identifiers that are keywords in C++ but not in C and then do pointer comparisons on the `IdentifierInfo*` maybe? At least then we wouldn’t have to do actual string comparisons; that said, I’m not sure that’s going to be that much better...

AaronBallman committed:
Handle alternative keywords

AaronBallman committed:
Rename flag

AaronBallman commented:
> > But the token kind we have is `identifier` and not one of the keywords, so we can't do the more trivial integer comparisons.
> 
> Can we precompute a list/map/whatever of all identifiers that are keywords in C++ but not in C and then do pointer comparisons on the `IdentifierInfo*` maybe? At least then we wouldn’t have to do actual string comparisons; that said, I’m not sure that’s going to be that much better...

That's a pretty good idea, let me try that. I'll back out the changes that disable checking when the diagnostic isn't enabled so we get a better comparison.

AaronBallman committed:
Temporarily back this change out

I want to see if I can improve the speed before we conditionally enable
the check.

AaronBallman committed:
Remove the need for string comparisons

If this ends up performing significantly better, I'll rip out the
changes to IdentifierTable.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

AaronBallman committed:
Remove old implementation, go back to checking only when enabled

AaronBallman commented:
> > > But the token kind we have is `identifier` and not one of the keywords, so we can't do the more trivial integer comparisons.
> > 
> > 
> > Can we precompute a list/map/whatever of all identifiers that are keywords in C++ but not in C and then do pointer comparisons on the `IdentifierInfo*` maybe? At least then we wouldn’t have to do actual string comparisons; that said, I’m not sure that’s going to be that much better...
> 
> That's a pretty good idea, let me try that. I'll back out the changes that disable checking when the diagnostic isn't enabled so we get a better comparison.

That idea clawed back most of the performance:
```
Initial slow-down: C 		56a3f3cd28 	8.04s (+2.96%) 	8.23s (+3.32%) 	11.87s (+1.92%) 	2.83s (+10.21%) 	8.50s (+2.99%) 	3.22s (+7.82%) 	7.12s (+1.63%) 	2.43s (+5.91%) 	580.64s (+0.44%)
New approach: C 		07e44e05f8 	7.80s (-2.85%) 	8.03s (-2.86%) 	11.63s (-1.57%) 	2.57s (-8.93%) 	8.25s (-2.60%) 	2.97s (-7.09%) 	7.06s (-1.02%) 	2.33s (-3.10%) 	580.01s (+0.01%)
```

erichkeane reviewed (commented):
sorry!  I made these comments a while ago, but apparently never submitted them.

AaronBallman commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -6107,6 +6109,29 @@ static bool isFromSystemHeader(SourceManager &SM, const Decl *D) {
+#include "clang/Basic/TokenKinds.def"
+  }
+
+  return Keywords.contains(II);
```
32 is wrong though, we have 34 keywords. Good catch on that. I'll bump to 48 to give wiggle room and pick something power-of-two-ish. WDYT?

erichkeane replied:
Oof!  48 would be fine.

Else, I'd love a static-assert (or, if you're doing the work for that, just making this the 'right' number so it never actually allocates and we don't have 'extra'), but can live with '48 should be big enough for anybody!'

erichkeane reviewed (approved):
You can remove the `Params.empty` if you'd like, but this LGTM anyway.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

AaronBallman committed:
Update based on review feedback

* Switch to using a binary_search
* Remove an unnecessary check for empty()

AaronBallman commented:
After talking with Erich offline, I've updated the implementation to use a binary_search. The pointers are being cast to `uintptr_t` so that the `llvm::sort()` isn't operating on unrelated pointers.

AaronBallman committed:
One more iteration

This one uses std::array so the size is exact.

efriedma-quic commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -6107,6 +6109,43 @@ static bool isFromSystemHeader(SourceManager &SM, const Decl *D) {
+      &S.getPreprocessor().getIdentifierTable().get(#NAME));
+#include "clang/Basic/TokenKinds.def"
+    assert(Idx == Keywords.size() && "expected to fill every member!");
+    llvm::sort(Keywords);
```
This isn't thread-safe.

Maybe make the "Keywords" array a member of Sema?

AaronBallman replied:
Wait... Clang is generally supposed to be thread safe? I thought only very specific interfaces in Clang were thread-safe as a purely best-effort mechanism because we don't test anything in a multi-threaded way? (Otherwise, what guards things like adding new nodes to the AST?)

AaronBallman replied:
Oh wait, this isn't really about thread safety in terms of race conditions. It's about being able to initialize clang as a library from multiple threads?

efriedma-quic replied:
Thread safety in the sense that we want to allow running multiple instances of the parser in the same process, yes.

Are IdentifierInfo pointers equal across different instances of the parser?  If they are... you can probably do something with a static constructor.  ManagedStatic is probably overkill here; just use a lambda in the initializer.

AaronBallman replied:
Okay, easy enough, thanks!

AaronBallman committed:
Change initialization; NFC

zygoloid commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -6107,6 +6109,44 @@ static bool isFromSystemHeader(SourceManager &SM, const Decl *D) {
+  // or not. Note, this treats all keywords as being enabled, regardless of the
+  // setting of other language options. It intentionally disables the modules
+  // keywords because those are conditional keywords, so may be safe to use.
+  static auto Keywords = [&S] {
+    std::array<uintptr_t, countCPlusPlusKeywords()> Ret;
+    unsigned Idx = 0;
+#define MODULES_KEYWORD(NAME)
+#define KEYWORD(NAME, FLAGS)                                                   \
+  Ret[Idx++] = reinterpret_cast<uint64_t>(                                     \
+      &S.getPreprocessor().getIdentifierTable().get(#NAME));
```
This will break for at least all users of Clang as a library -- we can't store pointers to `IdentifierInfo`s in a static local table, because we can have multiple preprocessors (and hence identifier tables and `IdentifierInfo`s for a single identifier) in a process.

zygoloid commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -6107,6 +6109,44 @@ static bool isFromSystemHeader(SourceManager &SM, const Decl *D) {
+    return Ret;
+  }();
+
+  return llvm::binary_search(Keywords, reinterpret_cast<uintptr_t>(II));
```
A binary search for each identifier still seems excessively expensive. Can we instead just store a flag on the `IdentifierInfo` that we set up when building the initial table? That's what we do in other similar cases (eg, `IsExtension` / `IsFutureCompatKeyword`); it doesn't seem like we need to invent something different here.

erichkeane replied:
Based on our count, this leaves us with ~6 compares at worst, and only when the flag was enabled.  According to benchmarking, it is sub-0.5% perf-hit when the warning is enabled, so expense didn't really show up as overly measurable I think?

As far as the flag on IdentifierInfo... I'm amazed to find out we have bits left to do so!  Apparently we have 22 bits left according to the comment/my count confirming it.

I'd considered suggesting it at one point but apparently miscounted us having bits left over (an off-by-10 count is.... something), and didn't want every IdentifierInfo to pay the cost.  

I think this is going to be the best way forward here.  A slight startup cost, to calculate these, but makes this a way easier patch to swallow.

AaronBallman replied:
Good suggestion, I've gone ahead with those changes.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

AaronBallman committed:
Go with another implementation of the keyword logic

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

zygoloid commented on clang/lib/Basic/IdentifierTable.cpp:
```diff
@@ -258,8 +283,13 @@ static void AddKeyword(StringRef Keyword,
-  if (AddResult == KS_Disabled) return;
+  // Don't add this keyword if disabled in this language and isn't otherwise
+  // special.
+  if (AddResult == KS_Disabled) {
```
Optional:

It'd be more consistent with the existing code to add an extra KeywordStatus value for this case, analogous to KS_Future. To do that, maybe we'd need to also replace the special macro for C++ operator keywords with a normal flag so that it can be taken into account when computing the KeywordStatus too.

I don't think it's necessary to be consistent here, though; we can always refactor later.

zygoloid commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -6118,6 +6120,10 @@ void Sema::warnOnReservedIdentifier(const NamedDecl *D) {
         << D << static_cast<int>(Status);
   }
+  // Diagnose use of C++ keywords in C as being incompatible with C++.
+  if (const IdentifierInfo *II = D->getIdentifier();
```
Is there a reason to warn on this here rather than in the same place we warn on "future keywords"? (When the preprocessor produces the final identifier token.) I would worry that this check would miss some cases. (Parameter names appear to need special handling, and they're probably not going to remain the only case even if they currently are.)

AaronBallman replied:
I was thinking this was similar to reserved identifiers, but I think your suggestion is cleaner. I'll work on an update for that

AaronBallman committed:
Implement in the preprocessor rather than Sema

AaronBallman committed:
Handle Objective-C @ keywords

AaronBallman commented:
Adding @rjmccall because this touches on Objective-C `@` keyword handling. (Note, maybe we want `-Wc++-compat` to be disabled in Objective-C entirely and we should have `-Wobjc++-compat` instead? `[foo bar];` is not valid C++, but I doubt anyone wants that diagnosed in Objective-C code, so it's a bit unclear what the diagnostic group means for ObjC.)

rjmccall commented:
The ObjC `@` is essentially an escape into a completely different grammar, and it doesn't matter whether the following identifier is a keyword or not in the base language. This warning should never kick in on that identifier.

Similarly, ObjC selector components exist outside of the normal keyword rules, and the warning should never kick in on them.

Otherwise, this warning would be very useful in baseline Objective-C as a way of discovering names that would be keywords in Objective-C++, mostly in the incorporated C subset but also for ObjC features that ultimately end up in the C namespace and therefore are subject to the normal keyword rules, like method parameters and class names.

AaronBallman committed:
Update for Objective-C again

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

AaronBallman committed:
Repair OpenMP test

AaronBallman commented:
> The ObjC `@` is essentially an escape into a completely different grammar, and it doesn't matter whether the following identifier is a keyword or not in the base language. This warning should never kick in on that identifier.

Okay, I've done that.

> Similarly, ObjC selector components exist outside of the normal keyword rules, and the warning should never kick in on them.

I don't know about selectors all that much; can you give me a test case that you think I should handle?

> Otherwise, this warning would be very useful in baseline Objective-C as a way of discovering names that would be keywords in Objective-C++, mostly in the incorporated C subset but also for ObjC features that ultimately end up in the C namespace and therefore are subject to the normal keyword rules, like method parameters and class names.

Okay, it remains enabled in Objective-C, just silenced if the preceding token is `@`.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

rjmccall commented:
> > Similarly, ObjC selector components exist outside of the normal keyword rules, and the warning should never kick in on them.
> 
> I don't know about selectors all that much; can you give me a test case that you think I should handle?

In general: look at `ParseObjCSelectorPiece` and how it generally accepts keywords in the selector. Actually, the list of keywords there seems out of date and notably doesn't include post-C++98 keywords like `decltype`. The logic is absolutely meant to encompass every keyword; note that it even includes basic C keywords like `if` and `struct`. You probably don't want to fix that, but I can ping someone at Apple to do it.

Anyway, selectors show up in three places in the ObjC grammar:
- `@selector` expressions, e.g. `@selector(virtual:typeid:bool:)`
- message send expressions, e.g. `[obj virtual: 0 typeid: 0 bool: 0]`
- method declarations, e.g. `
-  declarations, e.g.

```
@interface Foo
- (void) virtual: (int) x typeid: (int) y bool: (int) z;
@end
```

But there are also some special-case parsing rules for `class` because it is an important property name for `NSObject`. For example, in Objective-C++, the member name following `-> ` or `.` is allowed to be `class` despite it being a keyword.

AaronBallman commented:
> > > Similarly, ObjC selector components exist outside of the normal keyword rules, and the warning should never kick in on them.
> > 
> > 
> > I don't know about selectors all that much; can you give me a test case that you think I should handle?
> 
> In general: look at `ParseObjCSelectorPiece` and how it generally accepts keywords in the selector. Actually, the list of keywords there seems out of date and notably doesn't include post-C++98 keywords like `decltype`. The logic is absolutely meant to encompass every keyword; note that it even includes basic C keywords like `if` and `struct`. You probably don't want to fix that, but I can ping someone at Apple to do it.
> 
> Anyway, selectors show up in three places in the ObjC grammar:
> 
>     * `@selector` expressions, e.g. `@selector(virtual:typeid:bool:)`
> 
>     * message send expressions, e.g. `[obj virtual: 0 typeid: 0 bool: 0]`
> 
>     * method declarations, e.g. `
> 
>     * declarations, e.g.
> 
> 
> ```
> @interface Foo
> - (void) virtual: (int) x typeid: (int) y bool: (int) z;
> @end
> ```
> 
> But there are also some special-case parsing rules for `class` because it is an important property name for `NSObject`. For example, in Objective-C++, the member name following `-> ` or `.` is allowed to be `class` despite it being a keyword.

Ooof, it seems like there may not be a good way of handling this with the approach recommended by @zygoloid where we diagnose from `HandleIdentifier()` in the preprocessor. That requires far too much looking around to try to identify whether we're in the middle of a selector or not. The best I could think of would be to use some sort of "don't complain about keyword compatibility, we're doing <things>" RAII object to set a flag on the preprocessor so we can read it in `HandleIdentifier`. If we go that route, I could also stop looking to see if the preceding token is `@` from `HandleIdentifier()` and just use that mode. But is this too much of a hack?

rjmccall commented:
Yeah, triggering this at the preprocessor level is not going to let you do in properly in Objective-C, where keywords are effectively context-sensitive. If that's the only reasonable way to implement it, you should just turn it off in Objective-C completely.

zygoloid commented:
For what it's worth, the existing "keyword from language you're not using" warnings [already fire for analogous Objective-C++ cases](https://godbolt.org/z/3K6835PPK)... which in those cases is correct, since Clang does *not* accept C++11-onwards keywords as selector names in the corresponding versions of Objective-C++.

I wonder if it'd suffice to do a single-token lookahead / lookbehind in Objective-C / Objective-C++ and disable all of these warnings if the previous token is `@` or the next token is `:`. That would have false negatives for some weird cases like `class co_await : base {};` under `-Wc++20-compat` in Objective-C++17 or earlier, but that might be better than losing the warning entirely.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into perf/aballman-gh21898

AaronBallman committed:
Remove support for ObjC, update release note

AaronBallman commented:
I've removed support for Objective-C for now; we can come back to it again in the future.

rjmccall commented:
> For what it's worth, the existing "keyword from language you're not using" warnings [already fire for analogous Objective-C++ cases](https://godbolt.org/z/3K6835PPK)... which in those cases is correct, since Clang does _not_ accept C++11-onwards keywords as selector names in the corresponding versions of Objective-C++ (but presumably should -- I wonder if we should just check whether the token has an identifier spelling and not care what kind it is, [like we do for attribute names](https://github.com/llvm/llvm-project/blob/c3715ecb1e1e63856a029be3c95b7e9595a56018/clang/lib/Parse/ParseDeclCXX.cpp#L4575)?).

Yeah, I can't think of any good reason why this has a hardcoded list of keywords. Even if it's desirable for switching purposes, seems like it should be autogenerated from `TokenKinds.def`.

> I wonder if it'd suffice to do a single-token lookahead / lookbehind in Objective-C / Objective-C++ and disable all of these warnings if the previous token is `@` or the next token is `:`

Sadly, nullary selector pieces aren't followed by colons: `[foo decltype]`, etc.

> I've removed support for Objective-C for now; we can come back to it again in the future.

I think that's fine.

nikic commented on clang/lib/Basic/IdentifierTable.cpp:
```diff
@@ -250,6 +250,32 @@ static KeywordStatus getKeywordStatus(const LangOptions &LangOpts,
+      break; // Go to the next flag, try again.
+    }
+  }
+  return false;
```
I'm a bit confused by this function. Why does this go through all bits in a loop instead of directly checking them? I'd have expected this whole function to be something like `return Flags & (KEYALLCXX|BOOLSUPPORT|WCHARSUPPORT|CHAR8SUPPORT);`

AaronBallman replied:
Code was lifted from https://github.com/llvm/llvm-project/blob/a4ceac7e3e04c32cb3e334eb89b54d8e99a298f8/clang/lib/Basic/IdentifierTable.cpp#L240 but yeah, this seems to be something I can simplify

AaronBallman replied:
I updated in 0b2ab11321d2b400987e9ade28a221dff67aea7d thank you for the catch!
